lib/Index/Index.cpp

```diff
@@ -440,6 +451,43 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
     return true;
   }
 
+  /// This method identifies containing declarations. Any references that are
+  /// associated with these declarations will obtain the \c RelationContainedBy
+  /// role.
+  ///
+  /// The use of \c PatternBindingDecl here ensures that variable declarations
+  /// will contain references to explicit types, initializers, custom
+  /// attributes, etc.
+  ///
+  /// Note that any single entry into the \c ContainerStack may have multiple
+  /// declarations, such is the case for a tuple, e.g:
+  /// let (a, b): (Int, String) = (x, y)
+  /// Currently both \c a and \c b will contain references to \c Int, \c String,
+  /// \c x and \c y. Future work could improve upon this to try match up the
+  /// indices, such that \c a would only contain references to \c Int and \c x,
+  /// and \c b references to \c String and \c y.
```

Comment:
As noted here, I would like to address this in the future. However, after some initial investigation my gut tells me that this might be quite tricky to implement, and even if possible, perhaps too brittle to warrant the complexity. If anyone can think of a nice way to achieve this, I'd love to hear it.

Reply:
I'm concerned about this multiplicative factor.  Your tests show the quadratic behaviour in the number of decls:
```swift
let (tupleElementA, tupleElementB): (Int, String) = (intValue, stringValue)
```

We are going to get `O(n^2)` relations from this, where `n` is the number of decls.

But there is also the number of references in any of those initializer expressions:
```swift
let ... = (intValue + intValue + ...m... + intValue, ...)
```

So if `m` is the number of references in the longest initializer expression for any of those decls, we end up with `O(n^2 + n*m)` total.

Reply:
Yes that's definitely concerning, though my reasoning was that the vast majority of cases are going to have only a few tuple elements, and that the extra references wouldn't be a cause for concern. However in hindsight that's probably not a defensible argument. I'll take another look into applying the references to the correct elements. If that's not possible however, I can think of a few other solutions:

1. Don't do anything for tuples, keep the current behavior.
2. Use on the first element as the container 🥴 
3. Emit the `PatternBinding` into the index as an implicit definition to contain the declarations and references? I'm not sure if this would even be possible.

Can you think of any other solutions?

Reply:
I'd prefer not to do (2), which seems actively wrong.

(3) seems workable to me, although not as nice as actually getting the accurate VarDecl parent. Something like

```
PatternBindingDecl
VarDecl childOf PBD
Type containedBy PBD
"Reference from initializer expression" containedBy PBD
```

On the plus side, this is straightforward to implement (I think), and the distinction between containedBy PBD and containedBy VarDecl seems uninteresting most of the time - actually very few queries even use containedBy relations in my experience.  On the minus side, it's less precise about which vardecl things are attached to, and it introduces a weird declaration kind into the index. Basically it forces the client code to worry about this weird complication.

Reply:
Ah, I didn't realise you wanted the initializer to be contained by the VarDecl as well (though it makes sense for your use case in hindsight). I'd prefer (3) over adding the references to all decls in the PDB.

I don't think linking up to accurate VarDecls would be *too* bad. You could implement a custom walker for the PDB and Pattern in `IndexASTWalker` (see `Traversal::visitPatternBindingDecl` and `Traversal::visitTypedPattern`). Though it would be a little unfortunate that we would then have two similar-but-not-quite walks.

Another question is what we would want to do for something like `var (a, b) = foo()` though - should `foo` be contained by both `a` and `b` in that case? Or maybe we should add the PDB regardless and have `foo` contained by PDB:

```
var (a, b) = foo()
var (c, d) = (bar(), baz())
```
->
```
PDB1 for (a, b)
a childOf PDB1
b childOf PDB1
foo containedBy PDB

PDB2 for (c, d)
c childOf PDB2
d childOf PDB2
bar containedBy c
baz containedBy d
```

Reply:
After a little thought, I think (3) has the potential to cause a fair amount of breakage for downstream users. At least for Periphery, I'd need to make some changes to support this. I'd likely need to relate each var to every reference anyway, so we've just pushed the quadratic behavior down the line.

There's also the question of what `clang::index::SymbolInfo::Kind` the PBD definition should have. I guess it could either be a `Variable`, or `*Property` to match the `VarDecl`, but then we'd still need some way to communicate that it _isn't really_. The other option is adding a new type, which seems like a no-go.

So perhaps the only viable options are (1) or (4) - matching the references to the decls.

> Another question is what we would want to do for something like var (a, b) = foo() though - should foo be contained by both a and b in that case?

Yes, I believe so. I feels correct to me? I'm also not sure what an alternative would be in this case.